torchgeo/datasets/utils.py

```diff
@@ -758,3 +761,131 @@ def convert_poly_coords(
+
+    Args:
+        mask: input mask array from `DatasetReader.dataset_mask()`
+              or `DatasetReader.read_masks()`.
+              Can be 2D or 3D with values between 0 and 255.
+        threshold: pixel values >= threshold are considered valid.
+            This is needed when/if the mask is based on alpha channel.
+            Defaults to 1 as we assume pixels outside the valid footprint will
+            also have 0 alpha.
+
+    Returns:
+        Binary mask of dtype uint8 with 255 = valid pixels, 0 = invalid.
+        If input is 3D, masks are combined (OR) before thresholding.
+
+    .. versionadded:: 0.8
+    """
+    if mask.ndim == 3:
+        # Combine multi-band masks: pixel valid if valid in any band. This is reached
+        # when input mask comes from `read_masks()` of multi-band raster.
+        combined = np.any(mask >= threshold, axis=0)
+    else:
+        combined = mask >= threshold
+
+    binary_mask: np.typing.NDArray[np.uint8] = (combined.astype(np.uint8)) * 255
+    return binary_mask
+
+
+def calculate_valid_footprint_from_binary_mask(
+    mask: np.typing.NDArray[np.uint8],
+    transform: Affine,
+    raster_width: int,
+    raster_resolution_x: float,
+) -> Polygon | MultiPolygon:
+    """Calculates valid data footprint from a raster's masks.
+
+    Args:
+        mask: binary mask where 255 representing valid pixels
+            and 0 representing nodata pixels
+        transform: affine transform for the raster
+        raster_width: width of the raster
+        raster_resolution_x: resolution of the raster in the x direction (meters)
+
+    Returns:
+        A `Polygon` or `MultiPolygon` representing the valid data footprint of the raster
+
+    .. versionadded:: 0.8
+    """
+    # Close eventual holes within the raster that have area smaller than 500 pixels.
+    # Yields two bands, one all-zero representing nodata pixels,
+    # the other representing valid data
+    # To ensure hole size is smaller than raster size cap them at 0.2%.
+    # This value was found empirically as per
+    # https://rasterio.readthedocs.io/en/stable/topics/masks.html#writing-masks
+    max_hole_size = min(int(mask.size * 0.002), 800) or 1  # failsafe for 0
+    sieved_mask = sieve(mask, max_hole_size)
+
+    # Extract polygon for valid data values. Only interested in the valid-band.
+    # To support complex footprints we allow multiple such polygons and merge them
+    # to one multipolygon later
+    geoms = [g for g, v in shapes(sieved_mask, transform=transform) if v > 0]
+
+    # Create a Shapely Polygon object(s)
+    vector_footprint = MultiPolygon(
+        [Polygon(feature['coordinates'][0]) for feature in geoms]
+    )
+    # The resulting polygon is very staggered/pixelated,
+    # which could result in thousands of corners.
+    # The valid data footprints are usually very simple (3-5 corners)
+    # Simplifying each side to length 1/5 of the raster size (assumes crs is meters)
```

Comment:
> (assumes crs is meters)

This is not a good assumption

Reply:
Yeah, I noted it down since I thought it might not be. If the CRS (of the Raster!) is in degrees, the resolution is a tiny number, and this tolerance will be huge, resulting in a very simplified polygon. How often is that the case for rasters?